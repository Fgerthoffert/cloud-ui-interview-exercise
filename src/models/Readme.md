This directory contains the rematch reducers (https://github.com/rematch/rematch), typically one per view.
