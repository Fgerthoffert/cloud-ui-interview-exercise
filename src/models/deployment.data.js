const sourceData = {
  regionId: "us-east-1",
  id: "affd5eb6-7e03-57c9-8e70-fe964a0c90a7",
  name: "Cross-group uniform application",
  displayId: "affd5e",
  displayName: "Cross-group uniform application",
  isStopped: false,
  healthy: false,
  user: {
    id: 407920,
    level: "standard",
    isPremium: false
  },
  plan: {
    healthy: true,
    isActive: true,
    isPending: false,
    version: "1.7.2",
    instanceCapacity: 1024,
    instanceCount: 1,
    availabilityZones: 2,
    configurationSteps: [
      {
        time: "2017-01-09T02:28:18.014Z",
        ok: true,
        type: "step-starting",
        value: "resolve-node-type"
      },
      {
        time: "2017-01-09T02:28:18.018Z",
        ok: true,
        type: "step-completed",
        value: "resolve-node-type"
      }
    ]
  },
  master: {
    healthy: true,
    count: 1
  },
  shards: {
    healthy: true,
    count: {
      total: 3,
      available: 3,
      unavailable: 0
    }
  },
  instances: {
    healthy: true,
    count: {
      total: 3,
      notRunning: 0,
      running: 3
    }
  },
  snapshots: {
    healthy: false,
    count: {
      total: 97
    },
    latest: {
      name: "scheduled-1483666147-instance-0000000056",
      success: true,
      time: "2017-01-06T01:29:25.142Z"
    }
  },
  kibana: {
    enabled: false
  },
  monitoring: {
    enabled: false
  }
};

export default {
  state: {
    deployment: sourceData,
    deploymentId: sourceData.id
  },
  reducers: {
    setDeployment(state, payload) {
      return { ...state, deployment: payload };
    },
    setDeploymentId(state, payload) {
      return { ...state, deploymentId: payload };
    }
  },
  effects: {
    async initView() {}
  }
};
