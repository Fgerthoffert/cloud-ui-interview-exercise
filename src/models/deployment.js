import axios from "axios";

export default {
  state: {
    deployment: {},
    deploymentId: null
  },
  reducers: {
    setDeployment(state, payload) {
      return { ...state, deployment: payload };
    },
    setDeploymentId(state, payload) {
      return { ...state, deploymentId: payload };
    }
  },
  effects: {
    async initView(deploymentId) {
      // Record the requested deployment ID. Useful to display the error message in case the actual deployment is not found (i.e. if the link was bookmarked).
      this.setDeploymentId(deploymentId);
      this.setDeployment({});

      // Potentially we'd call a specific endpoint returning details just about this particular deployment ID.
      const deployments = await axios.get("/deployments.json");
      if (deployments.data !== undefined) {
        // Will return undefined if no deployment exist with this ID
        const deployment = deployments.data.record.find(
          dep => dep.id === deploymentId
        );
        this.setDeployment(deployment);
      }
    }
  }
};
