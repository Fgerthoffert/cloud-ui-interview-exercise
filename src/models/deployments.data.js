import { buildFacets } from "../utils/facets/";

const facetsConfig = [
  { key: "regionId", name: "Regions", type: "text" },
  {
    key: "plan.healthy",
    name: "Plans",
    type: "bool",
    true: "Healthy",
    false: "Unhealthy"
  },
  {
    key: "master.healthy",
    name: "Masters",
    type: "bool",
    true: "Healthy",
    false: "Unhealthy"
  },
  {
    key: "shards.healthy",
    name: "Shards",
    type: "bool",
    true: "Healthy",
    false: "Unhealthy"
  },
  {
    key: "instances.healthy",
    name: "Instances",
    type: "bool",
    true: "Healthy",
    false: "Unhealthy"
  },
  {
    key: "snapshots.healthy",
    name: "Snapshots",
    type: "bool",
    true: "Healthy",
    false: "Unhealthy"
  }
];

const sourceData = {
  isSearching: false,
  totalCount: 5100,
  record: [
    {
      regionId: "us-east-1",
      id: "affd5eb6-7e03-57c9-8e70-fe964a0c90a7",
      name: "Cross-group uniform application",
      displayId: "affd5e",
      displayName: "Cross-group uniform application",
      isStopped: false,
      healthy: false,
      user: {
        id: 407920,
        level: "standard",
        isPremium: false
      },
      plan: {
        healthy: true,
        isActive: true,
        isPending: false,
        version: "1.7.2",
        instanceCapacity: 1024,
        instanceCount: 1,
        availabilityZones: 2,
        configurationSteps: [
          {
            time: "2017-01-09T02:28:18.014Z",
            ok: true,
            type: "step-starting",
            value: "resolve-node-type"
          },
          {
            time: "2017-01-09T02:28:18.018Z",
            ok: true,
            type: "step-completed",
            value: "resolve-node-type"
          }
        ]
      },
      master: {
        healthy: true,
        count: 1
      },
      shards: {
        healthy: true,
        count: {
          total: 3,
          available: 3,
          unavailable: 0
        }
      },
      instances: {
        healthy: true,
        count: {
          total: 3,
          notRunning: 0,
          running: 3
        }
      },
      snapshots: {
        healthy: false,
        count: {
          total: 97
        },
        latest: {
          name: "scheduled-1483666147-instance-0000000056",
          success: true,
          time: "2017-01-06T01:29:25.142Z"
        }
      },
      kibana: {
        enabled: false
      },
      monitoring: {
        enabled: false
      }
    },
    {
      regionId: "us-east-1",
      id: "071d305d-993f-590f-81c3-a462294674fa",
      name: "Down-sized upward-trending hub",
      displayId: "071d30",
      displayName: "Down-sized upward-trending hub",
      isStopped: false,
      healthy: true,
      user: {
        id: 108182,
        level: "standard",
        isPremium: false
      },
      plan: {
        healthy: true,
        isActive: true,
        isPending: false,
        version: "1.7.2",
        instanceCapacity: 2048,
        instanceCount: 1,
        availabilityZones: 1,
        configurationSteps: [
          {
            time: "2016-11-15T11:23:45.242Z",
            ok: true,
            type: "step-starting",
            value: "plan-validator"
          },
          {
            time: "2016-11-15T11:23:45.249Z",
            ok: true,
            type: "step-completed",
            value: "plan-validator"
          },
          {
            time: "2016-11-15T11:23:45.250Z",
            ok: true,
            type: "step-starting",
            value: "validate-elasticsearch-plugin-versions"
          }
        ]
      },
      master: {
        healthy: true,
        count: 1
      },
      shards: {
        healthy: true,
        count: {
          total: 97,
          available: 97,
          unavailable: 0
        }
      },
      instances: {
        healthy: true,
        count: {
          total: 1,
          notRunning: 0,
          running: 1
        }
      },
      snapshots: {
        healthy: true,
        count: {
          total: 100
        },
        latest: {
          name: "scheduled-1483932071-instance-0000000023",
          success: true,
          time: "2017-01-09T03:23:46.759Z"
        }
      },
      kibana: {
        enabled: false
      },
      monitoring: {
        enabled: false
      }
    }
  ]
};

export default {
  state: {
    deployments: sourceData.record,
    deploymentsFiltered: sourceData.record,
    query: {},
    facets: buildFacets({}, sourceData.record, facetsConfig)
  },
  reducers: {
    setDeployments(state, payload) {
      return { ...state, deployments: payload };
    },
    setDeploymentsFiltered(state, payload) {
      return { ...state, deploymentsFiltered: payload };
    },
    setQuery(state, payload) {
      return { ...state, query: payload };
    },
    setFacets(state, payload) {
      return { ...state, facets: payload };
    }
  },
  effects: {
    async initView() {},
    async updateQuery() {},
    async updateFacets() {},
    async updateDeployments() {}
  }
};
