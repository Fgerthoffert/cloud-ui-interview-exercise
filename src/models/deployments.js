import axios from "axios";

export default {
  state: {
    deployments: []
  },
  reducers: {
    setDeployments(state, payload) {
      return { ...state, deployments: payload };
    }
  },
  effects: {
    async initView() {
      // We could be doing something much fancier, but here I'm just loading the JSON file into an array to be consumed by the deployments view.
      const deployments = await axios.get("/deployments.json");
      if (deployments.data !== undefined) {
        this.setDeployments(deployments.data.record);
      }
    }
  }
};
