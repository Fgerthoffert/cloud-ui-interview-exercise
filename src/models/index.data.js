// This simply populate a store with test data
export { default as deployments } from "./deployments.data.js";
export { default as deployment } from "./deployment.data.js";
