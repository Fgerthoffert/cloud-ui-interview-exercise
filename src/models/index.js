export { default as deployments } from "./deployments.js";
export { default as deployment } from "./deployment.js";
