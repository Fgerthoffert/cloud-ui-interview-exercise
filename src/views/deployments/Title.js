import React, { Component } from "react";
import { connect } from "react-redux";
import PropTypes from "prop-types";

import { EuiPageHeader, EuiPageHeaderSection, EuiTitle } from "@elastic/eui";

/*
 Displays the page title, this is an example of narrow-focused component aimed at taking a redux props and displaying it
*/
class Title extends Component {
  render() {
    const { deployments } = this.props;
    return (
      <EuiPageHeader>
        <EuiPageHeaderSection>
          <EuiTitle size="l">
            <h1>
              Deployments
              {deployments.length && (
                <React.Fragment> ({deployments.length})</React.Fragment>
              )}
            </h1>
          </EuiTitle>
        </EuiPageHeaderSection>
      </EuiPageHeader>
    );
  }
}

Title.propTypes = {
  deployments: PropTypes.array.isRequired
};

const mapState = state => ({
  deployments: state.deployments.deployments
});

export default connect(
  mapState,
  null
)(Title);
