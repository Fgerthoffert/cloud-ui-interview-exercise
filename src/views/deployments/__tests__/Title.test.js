import React, { Component } from "react";
import renderer from "react-test-renderer";

import { init } from "@rematch/core";
import { Provider } from "react-redux";

import * as models from "../../../models";
import Title from "../Title";

const store = init({
  models
});

test("Test the view title", () => {
  const component = renderer.create(
    <Provider store={store}>
      <Title />
    </Provider>
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
