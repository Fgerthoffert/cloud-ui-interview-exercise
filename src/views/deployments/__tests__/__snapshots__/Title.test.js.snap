// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test the view title 1`] = `
<div
  className="euiPageHeader euiPageHeader--responsive"
>
  <div
    className="euiPageHeaderSection"
  >
    <h1
      className="euiTitle euiTitle--large"
    >
      Deployments
      0
    </h1>
  </div>
</div>
`;
